### PR TITLE
osd: full-object read crc is mismatch, because truncate modify oi.si…

### PR DESCRIPTION
**Pull request by jiawd**
…ze and forget to clear data_digest

Fixes: https://tracker.ceph.com/issues/53240
Signed-off-by: jiawd <jiawendong@xtaotech.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [ ] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by jiawd** [comment]
in my nautilus:
The object oi.size should be 4194304, but it is actually 4063232.
The object data_digest is 0xffffffff, but read crc is 0x29d98db8.

-----------------------------------------------------------------
in this case:
client write 0~4M, osd received multiple ops [0~128k], [131072~128k], [262144~128k], ..., [4063232~128k]
when write complete, oi.size should be 4194304.
but other truncate will modify op's length and truncate oi.size.

Consider two abnormal cases:
1. write after truncate, need to do truncate first and then do write.
2. write before truncate, need to trim length first and then do write.]

-----------------------------------------------------------------
in case2:
if truncate is to 0,
write is [0~128k]
trim length from 128k to 0, write change to [0~0], do nothing, oi.size is 0, x1 = set_data_digest(crc32(-1)).
write is [128k~128k]
trim length from 128k to 0, write change to [128k~0], truncate oi.size to offset 128k, x2 = set_data_digest(crc32(x1)).
write is [256k~128k]
trim length from 128k to 0, write change to [256k~0], truncate oi.size to offset 256k, x3 = set_data_digest(crc32(x2)).
...
write is [4063232~128k]
trim length from 128k to 0, write change to [4063232~0], truncate oi.size to offset 4063232, xn = set_data_digest(crs32(xn-1))

Now, we can see oi.size is 4063232, and data_digest is 0xffffffff, because thelength of in_data of crc is 0 every time.

---------------------------------------------------------------------
in conclusion, There are two logical problems that need to be fixed:
1. when truncate in write, need clear data_digest and DIGEST flag
2. when write before truncate, need to trim length, when offset over than oi.size, don't truncate oi.size to offset.

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -6724,10 +6725,16 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
+	    if (seq && (seq > op.extent.truncate_seq)) {
+	      //do nothing
+	      //write arrived after truncate, we should not truncate to offset
+	    } else {
+	      t->truncate(
+	        soid, op.extent.offset);
+	      truncate_update_size_and_usage(ctx->delta_stats, oi,
+	                                     op.extent.offset);
+	      oi.clear_data_digest();
+	    }
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:6737]: I don't think this part is correct.  The call to t->truncate here I think has nothing to do with the delayed truncate behavior -- it's to logically zero pad the object out to the offset of the new write.

Also, the block immediately below this one should already be calling oi.clear_data_digest().

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -6700,13 +6700,14 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
 	      ctx->clean_regions.mark_data_region_dirty(op.extent.truncate_size, oi.size - op.extent.truncate_size);
-	      oi.clear_data_digest();
 	    }
 	    if (op.extent.truncate_size != oi.size) {
               truncate_update_size_and_usage(ctx->delta_stats,
                                              oi,
                                              op.extent.truncate_size);
 	    }
+	    //truncate modify oi.size, need clear old data_digest and DIGEST flag
+	    oi.clear_data_digest();
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:6710]: This is only necessary if op.extent.truncate_size != oi.size -- let's move this up 2 lines under truncate_update_size_and_usage.

Comment (athanatos) [reply src/osd/PrimaryLogPG.cc:6710]: ...in fact, it should immediately follow every invocation of truncate_update_size_and_usage -- let's just refactor it into truncate_update_size_and_usage.
